Fix a race condition when using `parametersOf`

sergeys-opera opened the pull request:
`Scope` was incorrectly using `ThreadLocal` leading to a race condition. The actual issue was not putting `ArrayDeque` in a thread-local variable but creating the thread-local variable itself. Several concurrent threads could have created several instances of the `parameterStack` thread-local variable and only the last one would have used.
The correct usage of `ThreadLocal` is to instantiate it directly and let the `ThreadLocal` machinery to deal with synchronization.

This commit also adds a test that was failing prior to the fix. I've also added `iosSimulatorArm64` to the `:core:koin-test-coroutines` to test the fix on iOS.

This change fixes #2305.

sergeys-opera committed:
Fix a race condition when using `parametersOf`

`Scope` was incorrectly using `ThreadLocal` leading to a race
condition. The actual issue was not putting `ArrayDeque` in a
thread-local variable but creating the thread-local variable itself.
Several concurrent threads could have created several instances
of the `parameterStack` thread-local variable and only the last one
would have used.
The correct usage of `ThreadLocal` is to instantiate it directly
and let the `ThreadLocal` machinery to deal with synchronization.

This commit also adds a test that was failing prior to the fix.
I've also added `iosSimulatorArm64` to the `:core:koin-test-coroutines`
to test the fix on iOS.

This change fixes #2305.

arnaudgiuliani commented on projects/core/koin-core/src/commonMain/kotlin/org/koin/core/scope/Scope.kt:
```diff
@@ -306,13 +304,16 @@ class Scope(
 
     private fun getOrCreateParameterStack(): ArrayDeque<ParametersHolder> {
-        return parameterStack?.get() ?: ArrayDeque<ParametersHolder>().let { parameterStack = ThreadLocal(); parameterStack?.set(it) ; it }
+        val oldStack = parameterStack.get()
```
You remove ThreadLocal capacity to set those parameters?

sergeys-opera replied:
I don't. If the `ThreadLocal` `parameterStack` holds a parameters list already, it's returned.
If it's not, a new list is created and the existing `parameterStack` is updated.
The code works similarly to how it worked before (albeit not a one liner, I found it confusing to read) with one difference - instead of creating a new instance of a `ThreadLocal` variable we only create a list instance and update the existing `ThreadLocal` variable. This is how `ThreadLocal` is supposed to work.

sergeys-opera committed:
Make parameterStack a val

It is not reset anymore, no need for it to be a var

sergeys-opera committed:
Clear the contents of parameterStack upon closing the scope

Not that it matters, as the scope is being closed and the memory
will be released anyway, but it's more accurate and robust to
call `remove` instead of clearing the list manually.

sergeys-opera commented on projects/core/koin-core/src/commonMain/kotlin/org/koin/core/scope/Scope.kt:
```diff
@@ -431,8 +432,7 @@ class Scope(
 
-        parameterStack?.get()?.clear()
-        parameterStack = null
+        parameterStack.get()?.clear()
```
Fixed here https://github.com/InsertKoinIO/koin/pull/2306/commits/b170ba8cd8ac869211483ae117954b3f4e6089dc

arnaudgiuliani commented:
I took your proposal in #2345, and ensure ThreadLocal is not used if no use of parameters. 

Thanks
